### ISSUE TITLE
Issues with layers/sampler.py and the no_repeat_ngram_size parameter

### ISSUE THREAD

**Issue by hch-baobei** [vllm-project/vllm#757]
I'm adding no_repeat_ngram_size parameter and I have some problem with sampler.py.
in  transformers/generation/utils.py , my understanding order is model infer, (no_repeat processing: optional), temperature processing, topk processing, softmax. But in vllm, I found it seems like model infer,  temperature processing, softmax, topk processing.
Can you tell me why the order is different here or give me some advice to add no_repeat_ngram_size?
transformers

```
          outputs = self(
                **model_inputs,
                return_dict=True,
                output_attentions=output_attentions,
                output_hidden_states=output_hidden_states,
            )
            if synced_gpus and this_peer_finished:
                continue  # don't waste resources running the code we don't need

            next_token_logits = outputs.logits[:, -1, :]

            # pre-process distribution
            next_token_scores = logits_processor(input_ids, next_token_logits)
            next_token_scores = logits_warper(input_ids, next_token_scores)

            probs = nn.functional.softmax(next_token_scores, dim=-1)
            next_tokens = torch.multinomial(probs, num_samples=1).squeeze(1)

```

vllm

```
         hidden_states = self.model(......)
        hidden_states = _prune_hidden_states(hidden_states, input_metadata)

        # Get the logits for the next tokens.
        logits = torch.matmul(hidden_states, embedding.t())
        if embedding_bias is not None:
            logits += embedding_bias
        logits = gather_from_tensor_model_parallel_region(logits)
        # Remove paddings in vocab (if any).
        logits = logits[:, :self.vocab_size]

        # Apply presence and frequency penalties.
        output_tokens = _get_output_tokens(input_metadata)
        assert len(output_tokens) == logits.shape[0]
        presence_penalties, frequency_penalties = _get_penalties(
            input_metadata)
        assert len(presence_penalties) == logits.shape[0]
        assert len(frequency_penalties) == logits.shape[0]
        logits = _apply_penalties(logits, output_tokens, presence_penalties,
                                  frequency_penalties, self.vocab_size)

        # Apply temperature scaling.
        temperatures = _get_temperatures(input_metadata)
        assert len(temperatures) == logits.shape[0]
        if any(t != 1.0 for t in temperatures):
            t = torch.tensor(temperatures,
                             dtype=logits.dtype,
                             device=logits.device)
            # Use in-place division to avoid creating a new tensor.
            logits.div_(t.unsqueeze(dim=1))

        # We use float32 for probabilities and log probabilities.
        # Compute the probabilities.
        probs = torch.softmax(logits, dim=-1, dtype=torch.float)
        # Compute the log probabilities (before applying top-p and top-k).
        logprobs = torch.log(probs)

        # Apply top-p and top-k truncation.
        top_ps, top_ks = _get_top_p_top_k(input_metadata, self.vocab_size)
        assert len(top_ps) == len(top_ks) == probs.shape[0]
        do_top_p = any(p < 1.0 - _SAMPLING_EPS for p in top_ps)
        do_top_k = any(k != self.vocab_size for k in top_ks)
        if do_top_p or do_top_k:
            probs = _apply_top_p_top_k(probs, top_ps, top_ks)tates = self.model(......)
        hidden_states = _prune_hidden_states(hidden_states, input_metadata)

        # Get the logits for the next tokens.
        logits = torch.matmul(hidden_states, embedding.t())
        if embedding_bias is not None:
            logits += embedding_bias
        logits = gather_from_tensor_model_parallel_region(logits)
        # Remove paddings in vocab (if any).
        logits = logits[:, :self.vocab_size]

        # Apply presence and frequency penalties.
        output_tokens = _get_output_tokens(input_metadata)
        assert len(output_tokens) == logits.shape[0]
        presence_penalties, frequency_penalties = _get_penalties(
            input_metadata)
        assert len(presence_penalties) == logits.shape[0]
        assert len(frequency_penalties) == logits.shape[0]
        logits = _apply_penalties(logits, output_tokens, presence_penalties,
                                  frequency_penalties, self.vocab_size)

        # Apply temperature scaling.
        temperatures = _get_temperatures(input_metadata)
        assert len(temperatures) == logits.shape[0]
        if any(t != 1.0 for t in temperatures):
            t = torch.tensor(temperatures,
                             dtype=logits.dtype,
                             device=logits.device)
            # Use in-place division to avoid creating a new tensor.
            logits.div_(t.unsqueeze(dim=1))

        # We use float32 for probabilities and log probabilities.
        # Compute the probabilities.
        probs = torch.softmax(logits, dim=-1, dtype=torch.float)
        # Compute the log probabilities (before applying top-p and top-k).
        logprobs = torch.log(probs)

        # Apply top-p and top-k truncation.
        top_ps, top_ks = _get_top_p_top_k(input_metadata, self.vocab_size)
        assert len(top_ps) == len(top_ks) == probs.shape[0]
        do_top_p = any(p < 1.0 - _SAMPLING_EPS for p in top_ps)
        do_top_k = any(k != self.vocab_size for k in top_ks)
        if do_top_p or do_top_k:
            probs = _apply_top_p_top_k(probs, top_ps, top_ks)
```

**Comment by colefranks** [comment]
fwiw it now looks like top_p/top_k is done before softmax https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/layers/sampler.py

**Comment by pspdada** [comment]
Hello, I noticed that you previously raised this question in the vllm library regarding the implementation of the no_repeat_ngram_size parameter . I found that this feature is still not included in the current version of the vllm library. I would like to discuss with you how this feature can be added, as I currently have a strong need for this parameter (https://github.com/vllm-project/vllm/issues/7842).

**Comment by Syhen** [comment]
Anyone need no_repeat_ngram_size can implement `NoRepeatNGramLogitsProcessor` and past to the logits_processors.
The main code is copied from HuggingFace implementation, and did some adaptation for vllm.

```python
import torch
import vllm
from transformers import LogitsProcessor
from transformers.generation.logits_process import _calc_banned_ngram_tokens
from vllm import LLM, SamplingParams
from vllm.lora.request import LoRARequest


class NoRepeatNGramLogitsProcessor(LogitsProcessor):
    def __init__(self, ngram_size: int):
        if not isinstance(ngram_size, int) or ngram_size <= 0:
            raise ValueError(f"`ngram_size` has to be a strictly positive integer, but is {ngram_size}")
        self.ngram_size = ngram_size

    def __call__(self, prompt_tokens_ids: tuple, past_tokens_ids: tuple, scores: torch.FloatTensor) -> torch.FloatTensor:
        """
        :ref: https://github.com/vllm-project/vllm/blob/911c8eb0000b1f9d1fef99ac9e209f83d801bd0a/vllm/model_executor/layers/logits_processor.py#L186
        """
        # score: [B, vocab_size]
        # input_ids: [B, cur_len]
        input_ids = prompt_tokens_ids + past_tokens_ids
        if len(input_ids) < self.ngram_size:
            return scores

        if len(scores.shape) == 1:
            scores = scores.reshape(1, -1)

        num_batch_hypotheses = scores.shape[0]
        input_ids = torch.LongTensor(input_ids).reshape(num_batch_hypotheses, -1)
        cur_len = input_ids.shape[-1]
        scores_processed = scores.clone()
        banned_batch_tokens = _calc_banned_ngram_tokens(self.ngram_size, input_ids, num_batch_hypotheses, cur_len)
        for i, banned_tokens in enumerate(banned_batch_tokens):
            scores_processed[i, banned_tokens] = -float("inf")

        return scores_processed


if __name__ == '__main__':
        no_repeat_ngram_size = 10
        
        logits_processors = []
        if no_repeat_ngram_size > 0:
            logits_processors.append(NoRepeatNGramLogitsProcessor(no_repeat_ngram_size))
        
        llm = LLM(
            model="<model_name_or_path>",
            tensor_parallel_size=2,
            gpu_memory_utilization=0.90,
            trust_remote_code=True,
            dtype="bfloat16",
            enforce_eager=True,
            device='cuda',
            disable_log_stats=True,
            enable_lora=False,
        )
        
        tokenizer = llm.get_tokenizer()
        tokenizer.padding_side = "left"
        
        prompt = "<your prompt>"
        
        responses = llm.generate(
            [prompt],
            vllm.SamplingParams(
                n=1,
                best_of=1,
                top_p=0.95,
                temperature=0.6,
                seed=100,
                skip_special_tokens=True,
                max_tokens=4096,
                logits_processors=logits_processors,
                logprobs=None,
                include_stop_str_in_output=False,
                repetition_penalty=1.10,
                top_k=50,
            ),
            use_tqdm=True
        )
        print(response)
```
